Make "time" a job prefix

krobelus opened the pull request:
In particular, this allows `true && time true`, or `true; and time true`.

time is valid only as job prefix, so `true | time true` could call `/bin/time`.
(I think this could be changed if anyone cares)

See discussion in #6442

## TODOs:
- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.md

faho commented on src/parse_execution.cpp:
```diff
@@ -125,8 +125,8 @@ tnode_t<g::plain_statement> parse_execution_context_t::infinite_recursive_statem
-    tnode_t<g::statement> statement = first_job.child<1>();
-    tnode_t<g::job_continuation> continuation = first_job.child<2>();
+    tnode_t<g::statement> statement = first_job.child<2>();
+    tnode_t<g::job_continuation> continuation = first_job.child<3>();
```
So these hardcoded child numbers seem quite suspect to me, but I never really got how our parser works.

This is because you have

```
[time] [command|builtin|...] therest
```

and so you're ignoring not just the `command|builtin` part, but also the `time` now? And time's separate because it can be stacked on top of the other decorators?

If so, the comment should be changed to indicate that.

krobelus replied:
> So these hardcoded child numbers seem quite suspect to me, but I never really got how our parser works.

`tnode_t` provides functions that know about the grammar at compile time,
they can be used to catch errors at compile time, like when accessing a child that is out of bounds.

These here are fine because the values for `tnode_t::child` are given an
explicit type, so this will most likely fail to compile if a grammar
change invalidates the child number. (It is unlikely that two child
nodes have the same type.)

I had another place where I just used `tnode_t::child<1>().tag()`
(without specifying the nodes type, fixed now) - that is of dangerous
because the child type could change without the compiler noticing.

> And time's separate because it can be stacked on top of the other decorators?
> If so, the comment should be changed to indicate that.

Yes; the comment above those lines was misleading.
When accessing the `job_t`, we ignore prefixing variable assingments and `time`.
Non-plain statements, like `not` and other decorators are ignored later; I moved the comment down.

faho commented:
>In particular, this allows true && time true, or true; and time true.

They should be tested then.

See tests/checks/time.fish for two tests I already added, since you'll have to regex-match the times and units.

faho commented:
Milestoning this for 3.1 because I'd want `time` to be at least mostly done if it is to be released then. Alternatively we could remove it for 3.1 and add it back later.

krobelus commented:
The most important thing here is to get the grammar right, the rest kind of writes itself.
If you spot some use of time that should be allowed but isn't, please shout.
It might just be ok now - I just changed the grammar now to allow both `time not true` and `not time true` (just like bash with `!`). This implementation is a little bit more complex because of this.

Also added some tests.
One thing that was (and is still) buggy is `time sleep 1&`.

mqudsi commented:
It was my intention to disallow the usage of `time` in a pipeline because it doesn't really make sense otherwise (imho), and users should be timing an entire job with `time` at the start of the job itself.

For oddities like `true && time false` or whatever which are "allowed" in implementation currently in `master`, I would prefer to special case `time` itself and catch that as a parser error - like our other builtins, if the user wants the "system" `time` they can use `true && /usr/bin/time false` or `true && command time false`, etc. so that it is very clear what is being run.

For `time sleep 1&`, I would similarly like to either update the parser to reject that as invalid grammar (e.g. "only foreground jobs may be timed") or failing that, refuse to emit `time` output. The alternative is to detect when the background job ends and print the output then, but currently we don't do that until a new prompt (or at least we didn't use to) so the time would be incorrect.

krobelus commented:
> For oddities like `true && time false` or whatever which are "allowed" in implementation currently in `master`, I would prefer to special case `time` itself and catch that as a parser error.

I believe the grammar change here is for the better because most users would expect `time true && sleep 1` to only measure `true` (like in bash). For compound commands we have `time begin; ...; end`.

> 
> For `time sleep 1&`, I would similarly like to either update the parser to reject that as invalid grammar (e.g. "only foreground jobs may be timed") or failing that, refuse to emit `time` output.

Yeah, I think we should reject `time foo &` in the parser for now.
Probably also `foo | time bar` for the reasons you stated.

mqudsi commented:
If we disregard bash compatibility (`time` is almost exclusively used interactively and one-off, so compatibility isn't a huge matter), does your take on any of the changes differ?

krobelus commented:
I like the convenience of running `make && time foo` (currently that needs something like `make && begin time foo; end`), but I'm happy to hear use cases where the current grammar is more convenient. The main difference is the precedence with `&&` and `||`.

mqudsi commented:
I was missing the "conditional set it and leave it" scenario, it's definitely reasonable to support something like `make && time ./benchmark`. I'm ok with changing it for the usability reasons you've mentioned but let me look the particulars of the implementation over a bit more thoroughly. 👍

mqudsi commented:
Merged. Sorry it took so long!
